apps/api/src/app/workflows-v2/usecases/build-test-data/build-workflow-test-data.usecase.ts

```diff
@@ -10,29 +15,102 @@ import {
 
   @InstrumentUsecase()
   async execute(command: WorkflowTestDataCommand): Promise<WorkflowTestDataResponseDto> {
-    const _workflowEntity: NotificationTemplateEntity = await this.fetchWorkflow(command);
-    const toSchema = buildToFieldSchema({ user: command.user, steps: _workflowEntity.steps });
-    const payloadSchema = parsePayloadSchema(_workflowEntity.payloadSchema, { safe: true });
-    const payloadSchemaMock =
-      payloadSchema && Object.keys(payloadSchema.properties || {}).length > 0 ? mockSchemaDefaults(payloadSchema) : {};
+    const workflow = await this.fetchWorkflow(command);
+    const toSchema = this.buildToFieldSchema({ user: command.user, steps: workflow.steps });
+    const payloadSchema = await this.resolvePayloadSchema(workflow, command);
+    const payloadSchemaMock = this.generatePayloadMock(payloadSchema);
 
     return {
       to: toSchema,
       payload: payloadSchemaMock,
     };
   }
 
+  @Instrument()
+  private async resolvePayloadSchema(
+    workflow: NotificationTemplateEntity,
+    command: WorkflowTestDataCommand
+  ): Promise<JSONSchemaDto> {
+    if (workflow.payloadSchema) {
+      return parsePayloadSchema(workflow.payloadSchema, { safe: true }) || {};
+    }
+
+    return this.generatePayloadSchema({
+      environmentId: command.user.environmentId,
+      organizationId: command.user.organizationId,
+      workflowId: workflow._id,
+    });
+  }
+
+  private generatePayloadMock(schema: JSONSchemaDto): Record<string, unknown> {
+    if (!schema?.properties || Object.keys(schema.properties).length === 0) {
+      return {};
+    }
+
+    return mockSchemaDefaults(schema);
+  }
+
+  @Instrument()
+  private async generatePayloadSchema({
+    environmentId,
+    organizationId,
+    workflowId,
+  }: {
+    environmentId: string;
+    organizationId: string;
+    workflowId: string;
+  }): Promise<JSONSchemaDto> {
+    const controlValues = await this.controlValuesRepository.find({
+      _environmentId: environmentId,
+      _organizationId: organizationId,
+      _workflowId: workflowId,
+      level: ControlValuesLevelEnum.STEP_CONTROLS,
+    });
+
+    if (!controlValues?.length) {
+      return {};
+    }
+
+    const templateVars = this.extractTemplateVariables(controlValues);
+    if (templateVars.length === 0) {
+      return {};
+    }
+
+    const variablesExample = pathsToObject(templateVars, {
+      valuePrefix: '{{',
+      valueSuffix: '}}',
+    }).payload as Record<string, unknown>;
+
+    return convertJsonToSchemaWithDefaults(variablesExample);
+  }
+
+  private extractTemplateVariables(controlValues: Array<{ controls: unknown | null }>): string[] {
+    const flattenedControls = controlValues
+      .map((item) => item.controls)
+      .filter((control): control is NonNullable<typeof control> => control != null);
+
+    const concatenatedControlValues = flattenedControls.map(flattenObjectValues).flat().join(' ');
+
+    return extractTemplateVars(concatenatedControlValues);
```

Comment:
after this week learning i know that this will throw an error if the control value has an invalid state because we use liquid js there.

i need to double-check this implementation.

Reply:
solved, we will extract valid and invalid separately.